src/Algebra/Properties/Semiring/Binomial.agda

```diff
@@ -0,0 +1,299 @@
+------------------------------------------------------------------------
+-- The Agda standard library
+--
+-- The Binomial Theorem for *-commuting elements in a Semiring
+------------------------------------------------------------------------
+
+{-# OPTIONS --cubical-compatible --safe #-}
+
+open import Algebra.Bundles using (Semiring)
+open import Data.Bool.Base using (true)
+open import Data.Nat.Base as Nat hiding (_+_; _*_; _^_)
+open import Data.Nat.Combinatorics
+  using (_C_; nCn≡1; nC1≡n; nCk+nC[k+1]≡[n+1]C[k+1])
+open import Data.Nat.Properties as Nat
+  using (<⇒<ᵇ; n<1+n; n∸n≡0; +-∸-assoc)
+open import Data.Fin.Base as Fin
+  using (Fin; zero; suc; toℕ; fromℕ; inject₁)
+open import Data.Fin.Patterns using (0F)
+open import Data.Fin.Properties as Fin
+  using (toℕ<n; toℕ-fromℕ; toℕ-inject₁)
+open import Data.Fin.Relation.Unary.Top
+  using (view; ‵fromℕ; ‵inject₁; view-fromℕ; view-inject₁)
+open import Function.Base using (_∘_)
+open import Relation.Binary.PropositionalEquality.Core as ≡
+  using (_≡_; _≢_; cong; module ≡-Reasoning)
+
+module Algebra.Properties.Semiring.Binomial {a ℓ} (S : Semiring a ℓ) where
+
+open Semiring S
+open import Algebra.Definitions _≈_
+open import Algebra.Structures _≈_
+open import Algebra.Properties.Semiring.Sum S
+open import Algebra.Properties.Semiring.Mult S
+open import Algebra.Properties.Semiring.Exp S
+import Relation.Binary.Reasoning.Setoid setoid as ≈-Reasoning
+
+
+------------------------------------------------------------------------
+-- The Binomial Theorem for *-commuting elements in a Semiring
+-- Freely adapted from PR #1287 by Maciej Piechotka (@uzytkownik)
+
+module _ (x y : Carrier) where
```

Comment:
While this is really clever structure to the file, and I like the names it generates, unfortunately it's incredibly difficult to work out what the parameterisation of each lemma actually is. For example the crucial commutativity assumption for `theorem` actually occurs 150 lines earlier. I think we need to flatten out most of the module structure....

Is that okay?

Reply:
Ouch. Probably (of course), but ... fiddly? I'll try to look at it over the weekend.

Reply:
I'll have a look at this now.